src/hotspot/share/nmt/vmatree.hpp

```diff
@@ -0,0 +1,178 @@
+// The set of points is stored in a balanced binary tree for efficient querying and updating.
+class VMATree {
+  friend class VMATreeTest;
+  // A position in memory.
+  using position = size_t;
+
+  class AddressComparator {
+  public:
+    static int cmp(position a, position b) {
+      if (a < b) return -1;
+      if (a == b) return 0;
+      if (a > b) return 1;
+      ShouldNotReachHere();
+    }
+  };
+
+public:
+  enum class StateType : uint8_t { Reserved, Committed, Released };
+
+  // Each point has some stack and a flag associated with it.
+  struct RegionData {
+    const NativeCallStackStorage::StackIndex stack_idx;
+    const MEMFLAGS flag;
+
+    RegionData() : stack_idx(), flag(mtNone) {}
+
+    RegionData(NativeCallStackStorage::StackIndex stack_idx, MEMFLAGS flag)
+    : stack_idx(stack_idx), flag(flag) {}
+
+    static bool equals(const RegionData& a, const RegionData& b) {
+      return a.flag == b.flag &&
+             NativeCallStackStorage::StackIndex::equals(a.stack_idx, b.stack_idx);
+    }
+  };
+
+  static const RegionData empty_regiondata;
+
+private:
+  struct IntervalState {
+  private:
+    // Store the type and flag as two bytes
+    uint8_t type_flag[2];
+    NativeCallStackStorage::StackIndex sidx;
+
+  public:
+    IntervalState() : type_flag{0,0}, sidx() {}
+    IntervalState(const StateType type, const RegionData data) {
+      assert(!(type == StateType::Released) || data.flag == mtNone, "Released type must have flag mtNone");
+      type_flag[0] = static_cast<uint8_t>(type);
+      type_flag[1] = static_cast<uint8_t>(data.flag);
+      sidx = data.stack_idx;
+    }
+
+    StateType type() const {
+      return static_cast<StateType>(type_flag[0]);
+    }
+
+    MEMFLAGS flag() const {
+      return static_cast<MEMFLAGS>(type_flag[1]);
+    }
+
+    RegionData regiondata() const {
+      return RegionData{sidx, flag()};
+    }
+
+    const NativeCallStackStorage::StackIndex stack() const {
+     return sidx;
+    }
+  };
+
+  // An IntervalChange indicates a change in state between two intervals. The incoming state
+  // is denoted by in, and the outgoing state is denoted by out.
+  struct IntervalChange {
+    IntervalState in;
+    IntervalState out;
+
+    bool is_noop() {
+      return in.type() == out.type() &&
+             RegionData::equals(in.regiondata(), out.regiondata());
+    }
+  };
+
+public:
+  using VMATreap = TreapCHeap<position, IntervalChange, AddressComparator>;
+  using TreapNode = VMATreap::TreapNode;
+
+private:
+  VMATreap _tree;
+
+  // AddressState saves the necessary information for performing online summary accounting.
+  struct AddressState {
+    position address;
+    IntervalChange state;
+
+    const IntervalState& out() const {
+      return state.out;
+    }
+
+    const IntervalState& in() const {
+      return state.in;
+    }
+  };
+
+public:
+  VMATree() : _tree() {}
+
+  struct SingleDiff {
+    int64_t reserve;
+    int64_t commit;
```

Comment:
Hm. We use ssize_t in many places for working with memory deltas, It works on all our platforms. And I don't see a good alternative here. int64 is not a replacement. To me, int64 feels like using void* for pointers, obfuscating intent. Its a memory size, I'd like therefore to see that in code. Apart from the obvious 64/32 bit problem.

One could probably use ptrdiff_t, but no pointers are involved, it seems awkward. 

One could use size_t+boolean tupels, and that may be the hypercorrect way. It would solve the problem of not being 

> Btw: Yes, I know, we can underflow or overflow the diff, but in practice no one will allocate 2**64 bytes, I am willing to take that risk.

An overflow would occur at 2**63. On 32-bit, on 2**31. So, on 32-bit you cannot express data sizes >= 2GB correctly. Which seems to me like a real limit. 

Maybe a size_t+boolean tupel is the right way to go.

Reply:
You can express large ranges with `int64_t`. There's no problem on 32-bit systems, just a bit of extra instructions emitted and some stack spillage. Also gah, I was sure that I wrote 63 and not 64 :-). Why not just typedef the `int64_t` as `delta` and call it a day?

>An overflow would occur at 263. On 32-bit, on 231. So, on 32-bit you cannot express data sizes >= 2GB correctly. Which seems to me like a real limit.

If anything, we should switch the `position` from `size_t` to `uint64_t` in `VMATree`.

Reply:
int64_t as delta_t would be a pragmatic compromise. On 32-bit, it allows to represent the full +-4gb. On 64-bit, nobody cares for deltas > 2^63.

On 32-bit, the only annoying part then is casting to/from size_t.